File: fiber.php
<?php

// Load the Rust library
$ffi = FFI::cdef(
    "unsigned int fib_rust(unsigned int n);",
    "./rust_lib/fib_lib/target/release/libfib_lib.so" // Path to the Rust library
);

// PHP Fibonacci function
function fib_php($n) {
    if ($n <= 1) {
        return $n;
    }
    return fib_php($n - 1) + fib_php($n - 2);
}

$n = 45;

// Benchmark PHP implementation
$start_php = microtime(true);
$result_php = fib_php($n);
$time_php = (microtime(true) - $start_php) * 1000; // Convert to milliseconds

// Benchmark Rust implementation via FFI
$start_rust = microtime(true);
$result_rust = $ffi->fib_rust($n);
$time_rust = (microtime(true) - $start_rust) * 1000; // Convert to milliseconds

// Displaying measurable data
echo "PHP Result: $result_php\n";
echo "Rust Result: $result_rust\n";

$time_php_rounded = round($time_php, 2);
$time_rust_rounded = round($time_rust, 2);
$speedup = round(($time_php / $time_rust) * 100, 2);

echo "\nComparison Summary:\n";
echo "PHP Execution Time: {$time_php_rounded} ms\n";
echo "Rust Execution Time: {$time_rust_rounded} ms\n";
echo "Rust is {$speedup}% faster than PHP for n={$n}.\n";

File: fiber.rb
require 'ffi'
require 'benchmark'

# Ruby Fibonacci
def fib_ruby(n)
  return n if n <= 1
  fib_ruby(n - 1) + fib_ruby(n - 2)
end

# Rust Fibonacci via FFI
module FibRust
  extend FFI::Library
  ffi_lib './rust_lib/fib_lib/target/release/libfib_lib.so' # Adjust path for your system
  attach_function :fib_rust, [:uint], :uint
end

n = 45
ruby_time = nil
rust_time = nil

Benchmark.bm(10) do |x|
  x.report("Ruby:") do
    ruby_time = Benchmark.realtime { fib_ruby(n) }
    puts "Result: #{fib_ruby(n)}"
  end

  x.report("Rust (FFI):") do
    rust_time = Benchmark.realtime { FibRust.fib_rust(n) }
    puts "Result: #{FibRust.fib_rust(n)}"
  end
end

# Displaying measurable data
ruby_ms = (ruby_time * 1000).round(2)
rust_ms = (rust_time * 1000).round(2)
speedup = ((ruby_time / rust_time) * 100).round(2)

puts "\nComparison Summary:"
puts "Ruby Execution Time: #{ruby_ms} ms"
puts "Rust Execution Time: #{rust_ms} ms"
puts "Rust is #{speedup}% faster than Ruby for n=#{n}."
